provers/plonk/src/prover.rs

```diff
@@ -341,33 +341,38 @@ where
     fn round_2(
         &self,
         witness: &Witness<F>,
-        common_preprocessed_input: &CommonPreprocessedInput<F>,
+        cpi: &CommonPreprocessedInput<F>,
         beta: FieldElement<F>,
         gamma: FieldElement<F>,
     ) -> Round2Result<F, CS::Commitment> {
-        let cpi = common_preprocessed_input;
-        let mut coefficients: Vec<FieldElement<F>> = vec![FieldElement::one()];
+        let mut z_evaluations: Vec<FieldElement<F>> = vec![FieldElement::one()];
         let (s1, s2, s3) = (&cpi.s1_lagrange, &cpi.s2_lagrange, &cpi.s3_lagrange);
 
         let k2 = &cpi.k1 * &cpi.k1;
 
         let lp = |w: &FieldElement<F>, eta: &FieldElement<F>| w + &beta * eta + &gamma;
 
+        let z_nums: Vec<_> = (0..&cpi.n - 1)
+            .map(|i| {
+                lp(&witness.a[i], &cpi.domain[i])
+                    * lp(&witness.b[i], &(&cpi.domain[i] * &cpi.k1))
```

Comment:
Maybe this could be more efficient just by just iterating over cpi.domain zipped with witness.a, witness.b, witness.c